### PR TITLE
Refactor the internals of `Store<T>`

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is an overdue refactoring and renaming of some internals of
the `Store` type in Wasmtime. The actual implementation of `Store<T>`
has evolved from the original implementation to the point where some of
the aspects of how things are structured no longer makes sense. There's
also always been a lot of unnecessary gymnastics when trying to get
access to various store pieces depending on where you are in `wasmtime`.

This refactoring aims to simplify all this and make the internals much
easier to read/write. The following changes were made:

* The `StoreOpaque<'_>` type is deleted, along with the `opaque()`
  method.

* The `StoreInnermost` type was renamed to `StoreOpaque`.
  `StoreOpaque<'_>` is dead. Long live `StoreOpaque`. This renaming
  and a few small tweaks means that this type now suffices for all
  consumers.

* The `AsContextMut` and `AsContext` traits are now implemented for
  `StoreInner<T>`.

These changes, while subtly small, help clean up a lot of the internals
of `wasmtime`. There's a lot less verbose `&mut
store.as_context_mut().opaque()` now. Additionally many methods can
simply start with `let store = store.as_context_mut().0;` and use things
internally. One of the nicer aspects of using references directly is
that the compiler automatically reborrows references as necessary
meaning there's lots of less manual reborrowing.

The main motivation for this change was actually somewhat roundabout
where I found that when `StoreOpaque<'_>` was being captured in closures
and iterators it's 3 pointers wide which is a lot of data to move
around. Now things capture over `&mut StoreOpaque` which is just one
nice and small pointer to move around. In any case though I've long
wanted to revisit the design of these internals to improve the
ergonomics. It's not expected that this change alone will really have
all that much impact on the performance of `wasmtime`.

Finally a doc comment was added to `store.rs` to try to explain all the
`Store`-related types since there are a nontrivial amount.

### CODE REVIEW

**Comment by alexcrichton** [comment]
It's worth pointing out that one refactoring I actually tried out was "what if we used `Pin<Box<StoreInner<T>>>`? I actually got pretty far with this but there's a few reasons that I didn't commit to it:

* Primarily, I don't think `Pin` is the right abstraction. My goal is to represent how when you have `Pin<&mut StoreInner<T>>` or `Pin<&mut StoreOpaque>` you have to be careful about accessing the contents and not overwrite the entire structure. I misunderstood the guarantees of `Pin`, though. The [`Pin::set` method](https://doc.rust-lang.org/stable/std/pin/struct.Pin.html#method.set) is safe and allows, at any time, overwriting the contents of the pinned pointer. That means that whatever gymnastics we go through to use `Pin` doesn't actually provide us any more safety since it's still possibly to safely cause bad behavior.

* Otherwise `Pin` was primarily pretty unergonomic. The main point against it was that, as we do today, we are required to manually reborrow `Pin` every time it's passed around and used twice. 

So basically the ergonomic drop plus not actually providing much benefit led me to the conclusion that we simply need to continue to be vigilant within Wasmtime to do the right thing. We do still need the `PhantomPinned`, though, I think. I'm at least somewhat sure that we can't be marking all our pointers as `noalias` with all the aliasing we actually have...

File: crates/wasmtime/src/store.rs [hunk 1]
```diff
@@ -79,14 +157,50 @@ pub struct Store<T> {
     inner: ManuallyDrop<Box<StoreInner<T>>>,
 }
 
+/// Internal contents of a `Store<T>` that live on the heap.
+///
+/// The members of this struct are those that need to be generic over `T`, the
+/// store's internal type storage. Otherwise all things that don't rely on `T`
+/// should go into `StoreOpaque`.
 pub struct StoreInner<T> {
```
Comment (bjorn3) [inline crates/wasmtime/src/store.rs:165]: Can this be `pub(crate)`?

**Review by pchickey — APPROVED** [review]
Thanks, the docs made this easy to understand

File: crates/wasmtime/src/store.rs (hunk 1 above)
Comment (alexcrichton) [reply crates/wasmtime/src/store.rs:165]: Er forgot to answer this before I merged, but I don't think so. There's a lot of `pub` stuff already and I frequently get compiler errors about pub-crate things being in pub-interfaces, so I prefer to not deal with any of that.
